azavea/raster-foundry
# Add datasource management UI

## Overview


This PR is a first pass at creating a user-interface for the management of datasources. This ballooned a bit so I am drawing the line here. A lot of this will need design work, so I don't want to devote any more time to this before that happens. This PR handles:
  - creating a path for a user to get to their datasources
  - listing a user's datasources (see next list)
  - UI editing default color-corrections for a datasource
  - UI for creating, updating, and deleting false-color composite presets

Things this PR does leave some things that need to be addressed:
  - connecting the front-end to the API (the datasource service)
  - the API support for filtering datasources by user
  - UI to create a new datasource (will probably be like the new project create workflow)
  - UI polish and validation (sliders for default color correction? intelligent band mapping ui, etc)

### Checklist

~- [ ] Styleguide updated, if necessary~
~- [ ] Swagger specification updated, if necessary~
~- [ ] Symlinks from new migrations present or corrected for any new migrations~


### Notes

There is no imported scene listing. You'll see a button for that but it doesn't go anywhere.

Various buttons intentionally don't do anything yet: 'Start an Import', 'Create a new Datasource', 'Create and new False Color Composite', 'Save', etc.

The editing of the JSON fields is built on top of a buffering concept. We create a buffer (copy of the JSON object) and make the changes to the buffer. If the user want to persist the changes, we replace the entities JSON field with the buffer. If editing is cancelled, we re-init the buffer to match the object's JSON.

### Demo

![screen shot 2017-03-17 at 10 35 36 am](https://cloud.githubusercontent.com/assets/2442245/24048024/f8999a78-0afd-11e7-9b82-2f3f3ea452af.png)
![screen shot 2017-03-17 at 10 36 09 am](https://cloud.githubusercontent.com/assets/2442245/24048028/f89bb740-0afd-11e7-9540-5fd3dff0b9f8.png)
![screen shot 2017-03-17 at 10 36 29 am](https://cloud.githubusercontent.com/assets/2442245/24048029/f89ff2ce-0afd-11e7-8fbc-861682d0bcb5.png)
![screen shot 2017-03-17 at 10 36 38 am](https://cloud.githubusercontent.com/assets/2442245/24048025/f899e4ba-0afd-11e7-8fa1-e7a9e035044d.png)
![screen shot 2017-03-17 at 10 37 06 am](https://cloud.githubusercontent.com/assets/2442245/24048027/f89add84-0afd-11e7-8e82-abb0bd0facd2.png)
![screen shot 2017-03-17 at 10 37 28 am](https://cloud.githubusercontent.com/assets/2442245/24048026/f89aa12a-0afd-11e7-880d-58693c05dc00.png)

## Testing Instructions

 * Click around the new imports page
 * Ensure you can see the list of datasources
 * Ensure pagination works for datasources (you can adjust the `pageSize` param on the `list.controller.js` file to be less than the number of datasources you have
 * Ensure you can get to the pages to edit any datasource
 * Check that you can get to the color correction edit and color composite edit
 * Changes to color correction of color composites should be able to be cancelled; make changes to these and hit cancel, navigate away, or anything else you can think of and verify that the changes are reset.
 
Connects #1164
